### ISSUE TITLE
Iteration order of hash keys doesn't appear random

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9382]
Migrated from [rt.perl.org#55952](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=55952) (status was 'resolved')

Searchable as RT55952$

**Comment by p5pRT** [comment]
### From jsailor@jesnetplus.com

#### Created by jsailor@jesnetplus.com



It doesn't seem that the iteration order of hash keys is random\. On 5\.8\.8\, I
created a hash with a bunch of key/value pairs\, and they are always returned by
the hash's iterator in the same order\. I'm not sure if the apparent
lack\-of\-randomness if actually evidence of a problem or not\.

Example&#8203;:

&nbsp;    jsailor@&#8203;gentoo ~ $ cat hi\.pl 
&nbsp;    \#\!/usr/bin/perl \-l
&nbsp;    map \{ $foo\{"key$\_"\} = "val$\_"  \} 1 \.\. 10;
&nbsp;    map \{ print "$\_ => $foo\{$\_\}"   \} keys %foo;
&nbsp;    jsailor@&#8203;gentoo ~ $ for \(\(i=0; i\<10; i\+\+\)\); do
&nbsp;                        PERL\_HASH\_SEED\_DEBUG=1 perl hi\.pl | md5sum
&nbsp;                       done | sort | uniq \-c
&nbsp;    HASH\_SEED = 4030767996
&nbsp;    HASH\_SEED = 4231742271
&nbsp;    HASH\_SEED = 2242964373
&nbsp;    HASH\_SEED = 1223985970
&nbsp;    HASH\_SEED = 3159642496
&nbsp;    HASH\_SEED = 3233708610
&nbsp;    HASH\_SEED = 1197677821
&nbsp;    HASH\_SEED = 3042627731
&nbsp;    HASH\_SEED = 3316032634
&nbsp;    HASH\_SEED = 3720177465
&nbsp;         10 b9610c6717785a4f9e26c9913d1bbb06  \-
&nbsp;    jsailor@&#8203;gentoo ~ $ 

This report if for 10 particular keys inserted in a particular order on a
particular build of Perl\, as returned by keys\, but the order is just as
constant \(for me\) for&#8203;:

 \- a few different other sets of keys
 \- iteration with each
 \- Perl on Debian 5\.8\.8\-7etch3\, Perl on OSX 10\.4

Thanks\!

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Gentoo at Mon Nov 19 16:17:34 EST 2007.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.19-gentoo-r5+bsdreiser2_1, archname=i686-linux-thread-multi
    uname='linux gentoo.grakker.jesnetplus.com 2.6.19-gentoo-r5+bsdreiser2_1 #4 wed sep 26 23:23:37 edt 2007 i686 amd athlon(tm) xp 2800+ authenticamd gnulinux '
    config_args='-des -Darchname=i686-linux-thread -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dcc=i686-pc-linux-gnu-gcc -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr -Dlocincpth=  -Doptimize=-march=athlon-xp -O2 -fomit-frame-pointer -pipe -Duselargefiles -Dd_semctl_semun -Dscriptdir=/usr/bin -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dinstallman1dir=/usr/share/man/man1 -Dinstallman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3pm -Dinc_version_list=5.8.0 5.8.0/i686-linux-thread-multi 5.8.2 5.8.2/i686-linux-thread-multi 5.8.4 5.8.4/i686-linux-thread-multi 5.8.5 5.8.5/i686-linux-thread-multi 5.8.6 5.8.6/i686-linux-thread-multi 5.8.7 5.8.7/i686-linux-thread-multi  -Dcf_by=Gentoo -Ud_csh -Dusenm -Dusethreads -Di_ndbm -Di_gdbm -Di_db'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='i686-pc-linux-gnu-gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-march=athlon-xp -O2 -fomit-frame-pointer -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement'
    ccversion='', gccversion='3.4.6 (Gentoo 3.4.6-r2, ssp-3.4.6-1.0, pie-8.7.10)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='i686-pc-linux-gnu-gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lpthread -lnsl -lndbm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/lib/perl5/vendor_perl/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl/5.8.5
    /usr/lib/perl5/vendor_perl/5.8.5/i686-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.7
    /usr/lib/perl5/vendor_perl/5.8.7/i686-linux-thread-multi
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/site_perl/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl/5.8.4
    /usr/lib/perl5/site_perl/5.8.4/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.5
    /usr/lib/perl5/site_perl/5.8.5/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.6
    /usr/lib/perl5/site_perl/5.8.6/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.7
    /usr/lib/perl5/site_perl/5.8.7/i686-linux-thread-multi
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/5.8.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/jsailor
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/3.4.6:/opt/blackdown-jdk-1.4.2.03/bin:/opt/blackdown-jdk-1.4.2.03/jre/bin:/usr/games/bin
    PERLIO=stdio
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @Whiteknight

The order is not "random" per say\, just "apparently random"\. There is a
definite order\, I believe it's related to the order that items are
inserted into the hash\. The point about it being "apparently random" is
that the items aren't sorted\, and the order that they are stored in is
subject to change in later versions\.

I don't think this is a bug\, and it can probably be resolved\.

\-\-Andrew Whitworth

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ap

\* Jonathan Sailor \<perlbug\-followup@&#8203;perl\.org> \[2008\-06\-18 00&#8203;:55\]&#8203;:
> On 5\.8\.8\, I created a hash with a bunch of key/value pairs\, and
> they are always returned by the hash's iterator in the same
> order\.

That is exactly as it should be\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Jun 18\, 2008 at 01&#8203;:24&#8203;:10AM \+0200\, Aristotle Pagaltzis wrote&#8203;:
> \* Jonathan Sailor \<perlbug\-followup@&#8203;perl\.org> \[2008\-06\-18 00&#8203;:55\]&#8203;:
> > On 5\.8\.8\, I created a hash with a bunch of key/value pairs\, and
> > they are always returned by the hash's iterator in the same
> > order\.
> 
> That is exactly as it should be\.

Except that the documentation in perlsec said otherwise until late last
year \(post 5\.10\.0's release\)\. Fixed with this change&#8203;:

http&#8203;://public\.activestate\.com/cgi\-bin/perlbrowse/p/32688

The documentation being incorrect may explain things\.

Nicholas Clark

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
